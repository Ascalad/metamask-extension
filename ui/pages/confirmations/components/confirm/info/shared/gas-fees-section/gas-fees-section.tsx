import { TransactionMeta } from '@metamask/transaction-controller';
// TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31860
// eslint-disable-next-line @typescript-eslint/naming-convention
import React, { useCallback, useState } from 'react';
import { EditGasModes } from '../../../../../../../../shared/constants/gas';
import { ConfirmInfoSection } from '../../../../../../../components/app/confirm/info/row/section';
import { useConfirmContext } from '../../../../../context/confirm';
// TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31860
// eslint-disable-next-line @typescript-eslint/naming-convention
import EditGasPopover from '../../../../edit-gas-popover';
import { useSupportsEIP1559 } from '../../hooks/useSupportsEIP1559';
import { GasFeesDetails } from '../gas-fees-details/gas-fees-details';

const LegacyTransactionGasModal = ({
  closeCustomizeGasPopover,
  transactionMeta,
}: {
  closeCustomizeGasPopover: () => void;
  transactionMeta: TransactionMeta;
}) => {
  return (
    <EditGasPopover
      onClose={closeCustomizeGasPopover}
      mode={EditGasModes.modifyInPlace}
      transaction={transactionMeta}
    />
  );
};

export const GasFeesSection = () => {
  const { currentConfirmation: transactionMeta } =
    useConfirmContext<TransactionMeta>();

  const [showCustomizeGasPopover, setShowCustomizeGasPopover] = useState(false);
  const closeCustomizeGasPopover = useCallback(
    () => setShowCustomizeGasPopover(false),
    [setShowCustomizeGasPopover],
  );

  const { supportsEIP1559 } = useSupportsEIP1559(transactionMeta);

  if (!transactionMeta?.txParams) {
    return null;
  }

  return (
    <ConfirmInfoSection data-testid="gas-fee-section">
      <GasFeesDetails setShowCustomizeGasPopover={setShowCustomizeGasPopover} />
      {!supportsEIP1559 && showCustomizeGasPopover && (
        <LegacyTransactionGasModal
          closeCustomizeGasPopover={closeCustomizeGasPopover}
          transactionMeta={transactionMeta}
        />
      )}
    </ConfirmInfoSection>
  );
};
