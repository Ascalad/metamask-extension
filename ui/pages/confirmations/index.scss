/** Please import your files in alphabetical order **/
@import 'confirm-approve/index';
@import 'confirm/index';
@import 'template-confirmation/template-confirmation';
@import 'send/gas-display/index';
@import 'send/send';
@import 'token-allowance/index';
@import './components/index'
