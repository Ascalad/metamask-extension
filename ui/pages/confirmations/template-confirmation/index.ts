export { TemplateConfirmation } from './template-confirmation';
