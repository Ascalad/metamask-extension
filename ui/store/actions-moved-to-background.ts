import {
  MetaMetricsEventOptions,
  MetaMetricsEventPayload,
  MetaMetricsPageOptions,
  MetaMetricsPagePayload,
} from '../../shared/constants/metametrics';
import {
  generateActionId,
  submitRequestToBackground,
} from './background-connection';

/**
 * @param payload - details of the page viewed
 * @param options - options for handling the page view
 */
export function trackMetaMetricsPage(
  payload: MetaMetricsPagePayload,
  options: MetaMetricsPageOptions,
) {
  return submitRequestToBackground('trackMetaMetricsPage', [
    { ...payload, actionId: generateActionId() },
    options,
  ]);
}

/**
 * @param payload - details of the event to track
 * @param options - options for routing/handling of event
 * @returns
 */
export function trackMetaMetricsEvent(
  payload: MetaMetricsEventPayload,
  options?: MetaMetricsEventOptions,
) {
  return submitRequestToBackground('trackMetaMetricsEvent', [
    { ...payload, actionId: generateActionId() },
    options,
  ]);
}
