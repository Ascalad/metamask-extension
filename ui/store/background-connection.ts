// eslint-disable-next-line import/no-restricted-paths
import { type MetaRpcClientFactory } from '../../app/scripts/lib/metaRPCClientFactory';

// TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31973
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Api = Record<string, (...params: any[]) => any>;
type BackgroundRpcClient = MetaRpcClientFactory<Api>;

const NO_BACKGROUND_CONNECTION_ERROR =
  'Background connection is not set. Please initialize the background connection before making requests.';

let background: BackgroundRpcClient | null = null;

export const generateActionId = () => Date.now() + Math.random();

/**
 * Promise-style call to background method invokes promisifiedBackground method directly.
 *
 * @param method - name of the background method
 * @param [args] - arguments to that method, if any
 * @returns
 */
export function submitRequestToBackground<R>(
  method: keyof Api,
  args?: Parameters<Api[typeof method]>,
): Promise<R> {
  if (!background) {
    throw new Error(NO_BACKGROUND_CONNECTION_ERROR);
  }
  return background[method](...(args ?? [])) as unknown as Promise<R>;
}

type CallbackMethod<R = unknown> = (error?: unknown, result?: R) => void;

/**
 * [Deprecated] Callback-style call to background method
 * invokes promisifiedBackground method directly.
 *
 * @deprecated Use async `submitRequestToBackground` function instead.
 * @param method - name of the background method
 * @param [args] - arguments to that method, if any
 * @param callback - Node style (error, result) callback for finishing the operation
 */
export const callBackgroundMethod = <R>(
  method: keyof Api,

  // TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31973
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  args: Parameters<Api[typeof method]>,
  callback: CallbackMethod<R>,
) => {
  submitRequestToBackground<R>(method, args).then(
    (result) => callback(null, result),
    callback,
  );
};

/**
 * Sets/replaces the background connection reference
 *
 * @param backgroundConnection
 */
export async function setBackgroundConnection(
  backgroundConnection: BackgroundRpcClient,
) {
  background = backgroundConnection;
}
