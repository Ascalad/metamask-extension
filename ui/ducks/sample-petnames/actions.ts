import type { Hex } from '@metamask/utils';
import type { MetaMaskReduxDispatch } from '../../store/store';
import { forceUpdateMetamaskState } from '../../store/actions';
import { submitRequestToBackground } from '../../store/background-connection';

export const ASSIGN_PET_NAME_METHOD = 'assignPetname';

export const assignPetname = (chainId: Hex, address: Hex, name: string) => {
  return async (dispatch: MetaMaskReduxDispatch) => {
    await submitRequestToBackground(ASSIGN_PET_NAME_METHOD, [
      chainId,
      address,
      name,
    ]);
    await forceUpdateMetamaskState(dispatch);
  };
};
