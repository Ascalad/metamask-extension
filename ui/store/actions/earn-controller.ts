import { submitRequestToBackground } from '../background-connection';

type refreshPooledStakingEligibilityOptions = {
  address?: string;
};

export async function refreshPooledStakingEligibility({
  address,
}: refreshPooledStakingEligibilityOptions = {}) {
  return await submitRequestToBackground<void>('refreshStakingEligibility', [
    { address },
  ]);
}
