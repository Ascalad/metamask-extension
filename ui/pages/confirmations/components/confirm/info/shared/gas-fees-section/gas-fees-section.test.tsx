import { act } from '@testing-library/react';
// TODO: Fix in follow-up ticket https://github.com/MetaMask/metamask-extension/issues/31860
// eslint-disable-next-line @typescript-eslint/naming-convention
import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import {
  getMockConfirmState,
  getMockContractInteractionConfirmState,
} from '../../../../../../../../test/data/confirmations/helper';
import { renderWithConfirmContextProvider } from '../../../../../../../../test/lib/confirmations/render-helpers';
import { getGasFeeTimeEstimate } from '../../../../../../../store/actions';
import { GasFeesSection } from './gas-fees-section';

jest.mock('../../../../../../../store/actions', () => ({
  ...jest.requireActual('../../../../../../../store/actions'),
  getGasFeeTimeEstimate: jest.fn(),
}));

jest.mock(
  '../../../../../../../components/app/alert-system/contexts/alertMetricsContext',
  () => ({
    useAlertMetrics: jest.fn(() => ({
      trackAlertMetrics: jest.fn(),
    })),
  }),
);

describe('<GasFeesSection />', () => {
  const middleware = [thunk];

  it('does not render component for gas fees section', () => {
    const state = getMockConfirmState();
    const mockStore = configureMockStore(middleware)(state);
    const { container } = renderWithConfirmContextProvider(
      <GasFeesSection />,
      mockStore,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders component for gas fees section', async () => {
    (getGasFeeTimeEstimate as jest.Mock).mockImplementation(async () =>
      Promise.resolve({ upperTimeBound: '1000' }),
    );

    const state = getMockContractInteractionConfirmState();
    const mockStore = configureMockStore(middleware)(state);
    let container;
    await act(async () => {
      const renderResult = renderWithConfirmContextProvider(
        <GasFeesSection />,
        mockStore,
      );
      container = renderResult.container;

      // Wait for any asynchronous operations to complete
      // TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31888
      // eslint-disable-next-line no-restricted-globals
      await new Promise(setImmediate);
    });

    expect(container).toMatchSnapshot();
  });
});
