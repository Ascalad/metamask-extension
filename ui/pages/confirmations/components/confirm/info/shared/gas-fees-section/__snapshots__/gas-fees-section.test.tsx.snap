// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GasFeesSection /> does not render component for gas fees section 1`] = `<div />`;

exports[`<GasFeesSection /> renders component for gas fees section 1`] = `<div />`;
